mono/metadata/threads.c

```diff
@@ -1839,16 +1841,15 @@ ves_icall_System_Threading_Thread_Join_internal (MonoThread *this_obj, int ms)
 	if (ret == MONO_THREAD_INFO_WAIT_RET_SUCCESS_0) {
 		THREAD_DEBUG (g_message ("%s: join successful", __func__));
 
-#ifdef HOST_WIN32
-		/* TODO: Do this on Unix platforms as well. See PR #5454 for context.  */
+		MonoNativeThreadId tid = thread_get_tid (thread);
```

Comment:
FYI, when `thread_get_tid()` was called here in a previous attempt to fix this issue it occasionally returned 0 on Windows. That is why I had to move the `thread_get_tid()` call up to the `LOCK_THREAD` block. Let me know and I can help you test your PR on Windows, just to make sure it works 100% of the time.

Reply:
Calling it earlier is not going to work since the thread might not been started yet, so it doesn't have a tid.

Reply:
Isn't that tested for here: https://github.com/vargaz/mono/blob/74c3322958194d74d3d47473230018965fc4409f/mono/metadata/threads.c#L1820? IIUC if the thread is unstarted this condition will trigger and the fuction returns `FALSE`?

Reply:
I think that test is racy, since the flag is not cleared in the newly created thread but in the creating thread. Obtaining the tid after the join () call should be better because join () will only return when the thread is close to terminating, which ensures it set its tid field.

Reply:
Did it return 0 on windows because of:
https://github.com/mono/mono/commit/96ace09a50ba9b93a6b88da2f6ed2c87b7381b99

Reply:
Aha, I wasn't aware of 96ace09. That could be why `tid` sometimes was 0. I will try your PR on the flaky `WaitHandleTest` on Windows today and let you know. The PR builds don't reliably test for this flakiness since it only happens occasionally.

Reply:
Ok, I have run `WaitHandleTest` 100 times on Windows x64 without seeing the previous flakiness. So as far as Windows is concerned this PR looks good!